File: engine/src/main/java/top/skyeyefast/mchjong/engine/ReplayMatch.java
package top.skyeyefast.mchjong.engine;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

/** Contains completed hands only. No RNG seed, future wall, or active hand is exported. */
public record ReplayMatch(UUID id, UUID tableId, long startedAt, long updatedAt, RuleConfig rules,
                          int initialDealer, List<Participant> participants, List<ReplayHand> hands, boolean complete, RedFives redFives) {
    public record Participant(UUID id, String name, boolean bot) {
        public Participant {
            Objects.requireNonNull(id); Objects.requireNonNull(name);
            if (name.isBlank() || name.length() > 128) throw new IllegalArgumentException("Invalid replay player name");
        }
    }
    public record Header(UUID id, long updatedAt, RuleConfig rules, int hands, boolean complete, List<String> names) {
        public Header {
            Objects.requireNonNull(id); Objects.requireNonNull(rules); names = List.copyOf(names);
            if (hands < 1 || hands > 1024 || names.size() != rules.players()
                || names.stream().anyMatch(name -> name.isBlank() || name.length() > 128))
                throw new IllegalArgumentException("Invalid replay header");
        }
    }
    public record Index(int page, String search, boolean oldestFirst, List<Header> matches, boolean more) {
        public Index {
            Objects.requireNonNull(search);
            matches = List.copyOf(matches);
            if (page < 0 || page > 100_000 || search.length() > 80 || matches.size() > 12)
                throw new IllegalArgumentException("Invalid replay page");
        }
    }

    public ReplayMatch {
        Objects.requireNonNull(id); Objects.requireNonNull(tableId); Objects.requireNonNull(rules);
        if (!rules.allows(redFives)) throw new IllegalArgumentException("Invalid replay red-five composition");
        participants = List.copyOf(participants); hands = List.copyOf(hands);
        if (participants.size() != rules.players() || initialDealer < 0 || initialDealer >= participants.size()
            || hands.size() > 1024 || complete && hands.isEmpty()) throw new IllegalArgumentException("Invalid replay match");
        for (var hand : hands) if (hand.initialHands().size() != rules.players()) throw new IllegalArgumentException("Replay rules mismatch");
        if (complete && (hands.getLast().finalScores().size() != rules.players() || hands.getLast().finalRanks().size() != rules.players()))
            throw new IllegalArgumentException("Missing final standings");
    }

    public Header header() {
        return new Header(id, updatedAt, rules, hands.size(), complete, participants.stream().map(Participant::name).toList());
    }
    public boolean permits(UUID player) {
        return participants.stream().anyMatch(participant -> !participant.bot() && participant.id().equals(player));
    }
    ReplayMatch append(ReplayHand hand, boolean ended) {
        var completed = new ArrayList<>(hands);
        completed.add(hand);
        return new ReplayMatch(id, tableId, startedAt, System.currentTimeMillis(), rules, initialDealer, participants, completed, ended, redFives);
    }
}

File: engine/src/main/kotlin/top/skyeyefast/mchjong/engine/ReplayMatch.kt
package top.skyeyefast.mchjong.engine

import java.util.UUID

/** Contains completed hands only. No RNG seed, future wall, or active hand is exported. */
@JvmRecord
data class ReplayMatch(
    val id: UUID,
    val tableId: UUID,
    val startedAt: Long,
    val updatedAt: Long,
    val rules: RuleConfig,
    val initialDealer: Int,
    val participants: List<Participant>,
    val hands: List<ReplayHand>,
    val complete: Boolean,
    val redFives: RedFives,
) {
    @JvmRecord
    data class Participant(val id: UUID, val name: String, val bot: Boolean) {
        init {
            require(name.isNotBlank() && name.length <= 128) { "Invalid replay player name" }
        }
    }

    @JvmRecord
    data class Header(
        val id: UUID,
        val startedAt: Long,
        val updatedAt: Long,
        val rules: RuleConfig,
        val hands: Int,
        val complete: Boolean,
        val names: List<String>,
        val finalScores: List<Double>,
        val finalRanks: List<Int>,
    ) {
        init {
            require(
                startedAt > 0 && updatedAt >= startedAt && hands in 1..1024 && names.size == rules.players() &&
                    names.none { it.isBlank() || it.length > 128 } &&
                    (finalScores.isEmpty() || finalScores.size == names.size) &&
                    (finalRanks.isEmpty() || finalRanks.size == names.size) &&
                    finalScores.none { !it.isFinite() } && finalRanks.none { it !in 1..names.size } &&
                    (!complete || finalScores.size == names.size && finalRanks.size == names.size),
            ) {
                "Invalid replay header"
            }
        }
    }

    @JvmRecord
    data class Index(
        val page: Int,
        val search: String,
        val oldestFirst: Boolean,
        val matches: List<Header>,
        val more: Boolean,
    ) {
        init {
            require(page in 0..100_000 && search.length <= 80 && matches.size <= 12) { "Invalid replay page" }
        }
    }

    init {
        require(rules.allows(redFives)) { "Invalid replay red-five composition" }
        require(
            participants.size == rules.players() && initialDealer in participants.indices && hands.size <= 1024 &&
                (!complete || hands.isNotEmpty()),
        ) { "Invalid replay match" }
        require(hands.all { it.initialHands.size == rules.players() }) { "Replay rules mismatch" }
        if (complete) {
            require(hands.last().finalScores.size == rules.players() && hands.last().finalRanks.size == rules.players()) {
                "Missing final standings"
            }
        }
    }

    fun header(): Header {
        val last = hands.last()
        return Header(
            id,
            startedAt,
            updatedAt,
            rules,
            hands.size,
            complete,
            java.util.List.copyOf(participants.map { it.name }),
            if (complete) java.util.List.copyOf(last.finalScores) else emptyList(),
            if (complete) java.util.List.copyOf(last.finalRanks) else emptyList(),
        )
    }

    fun permits(player: UUID): Boolean = participants.any { !it.bot && it.id == player }

    fun append(hand: ReplayHand, ended: Boolean): ReplayMatch =
        ReplayMatch(
            id,
            tableId,
            startedAt,
            System.currentTimeMillis(),
            rules,
            initialDealer,
            participants,
            java.util.List.copyOf(hands + hand),
            ended,
            redFives,
        )
}
